### PR TITLE
Deprecate ImmediateGeometry

### PR DESCRIPTION
**Pull request by reduz**
* Removed entirely from RenderingServer.
* Replaced by ImmediateMesh resource.
* ImmediateMesh replaces ImmediateGeometry, but could use more optimization in the future.
* Sprite3D and AnimatedSprite3D work again, ported from Godot 3.x (though a lot of work was needed to adapt them to Godot 4).
* RootMotionView works again.
* Polygon3D editor works again.

image
image
image

### CODE REVIEW

File: scene/resources/material.cpp [hunk 1]
```diff
@@ -2077,7 +2077,7 @@ BaseMaterial3D::TextureChannel BaseMaterial3D::get_refraction_texture_channel()
 	return refraction_texture_channel;
 }
 
-RID BaseMaterial3D::get_material_rid_for_2d(bool p_shaded, bool p_transparent, bool p_double_sided, bool p_cut_alpha, bool p_opaque_prepass, bool p_billboard, bool p_billboard_y) {
+Ref<Material> BaseMaterial3D::get_material_rid_for_2d(bool p_shaded, bool p_transparent, bool p_double_sided, bool p_cut_alpha, bool p_opaque_prepass, bool p_billboard, bool p_billboard_y, RID *r_shader_rid) {
```
Comment (JFonS) [inline scene/resources/material.cpp:2080]: Should this be renamed if it doesn't return an `RID` anymore?

File: scene/3d/sprite_3d.cpp [hunk 2]
```diff
@@ -34,15 +34,13 @@
 #include "scene/scene_string_names.h"
 
 Color SpriteBase3D::_get_color_accum() {
-	if (!color_dirty) {
```
Comment (JFonS) [inline scene/3d/sprite_3d.cpp:37]: Both `sprite_3d.cpp` and `sprite_3d.h` have a bunch of code style changes, like removing braces, changing `nullptr` to `NULL` and removing initialization from the header. I guess you used an old version of the file as a base?

**Review by JFonS — COMMENTED** [review]
Looks mostly good from an overall review. Just noticed some code-style regressions and a possible function rename.

File: scene/3d/sprite_3d.h [hunk 3]
```diff
@@ -176,8 +188,8 @@ class Sprite3D : public SpriteBase3D {
 	void set_frame(int p_frame);
 	int get_frame() const;
 
-	void set_frame_coords(const Vector2i &p_coord);
-	Vector2i get_frame_coords() const;
+	void set_frame_coords(const Vector2 &p_coord);
+	Vector2 get_frame_coords() const;
```
Comment (clayjohn) [inline scene/3d/sprite_3d.h:192]: This looks like it may be a regression

File: scene/3d/sprite_3d.cpp [hunk 4]
```diff
@@ -377,9 +444,8 @@ void Sprite3D::_draw() {
 	frame_offset *= frame_size;
 
 	Point2 dest_offset = get_offset();
-	if (is_centered()) {
+	if (is_centered())
```
Comment (JFonS) [inline scene/3d/sprite_3d.cpp:447]: Removed braces here :)

**Review by JFonS — APPROVED** [review]

**Comment by reduz** [comment]
:partying_face:

**Comment by akien-mga** [comment]
Thanks!

**Comment by MrJoshBowman** [comment]
Trying to test this but can't seem to be able to attach a script as it says "Script inherits from native type 'ImmediateMesh', so it can't be instantiated in object of type 'Node'.

I understand that it's the wrong object type, but I can't find what object it should actually be in the 'Change Node Type' list.

**Comment by Calinou** [comment]
ImmediateMesh is a *resource*, not a node.

**Pull request merged**